### ISSUE TITLE
CodegenConfigurator systemProperties are reused between code generation invokations

### ISSUE THREAD

**Issue by JLLeitschuh** [swagger-api/swagger-codegen#4788]
I'm using the swagger `CodegenConfigurator` in a gradle build where the swagger codegen project is loaded as a library and is being directly passed into the `DefaultGenerator`.

This usage in [this gradle plugin is very similar to my usage](https://github.com/thebignet/swagger-codegen-gradle-plugin/blob/master/src/main/groovy/org/detoeuf/SwaggerCodeGenTask.groovy#L42).

If I configure one `CodegenConfigurator` object with `systemProperties` being set, these settings seem to be being stored somewhere statically.

I'm observing that if I use the `DefaultGenerator` again in the same JVM I end up having files not get generated because of input from a previous run of the generator.

I know that the two `CodegenConfigurator` objects that I'm passing in are different because I've tested logging the `systemProperties` variable just before passing it in and for the first call it is populated and the second time it is empty.

**Comment by JLLeitschuh** [comment]
It seems that these lines are the one to blame in `CodegenConfigurator `:

```
    private void setSystemProperties() {
        for (Map.Entry<String, String> entry : systemProperties.entrySet()) {
            System.setProperty(entry.getKey(), entry.getValue());
        }
    }
```
https://github.com/swagger-api/swagger-codegen/blob/029728d851e63dc2887e126e1944f61a04d4540a/modules/swagger-codegen/src/main/java/io/swagger/codegen/config/CodegenConfigurator.java#L466

**Comment by otrosien** [comment]
@JLLeitschuh are you working on that subject?

**Comment by ePaul** [comment]
This is a general problem of using Java system properties – a kind of global variables – for this thing.
We should try to avoid this in general, and pass the parameters over other channels.

**Comment by otrosien** [comment]
@ePaul agree, do you have an idea? Beware that current users are relying on this feature.

**Comment by ePaul** [comment]
>  Beware that current users are relying on this feature.

On which part of the feature do they rely? Do they actually need system properties being set (like `file.separator`), or just need a way to pass custom parameters to parts of the code?

I guess we need to clean up all locations in the code which are using `System.getProperty()` to use other ways of passing parameters instead (like the additionalProperties map, or maybe even have a "systemProperties" map being passed to the Codegens). The Configurator should put the data from the `systemProperties` parameter into the new map instead of setting system properties.

Paul

**Comment by JLLeitschuh** [comment]
I'm not. I just use the ugly synchronization hack you've seen in my codegen plugin.

**Comment by otrosien** [comment]
@ePaul do you want to go ahead and tackle this issue?

**Comment by ePaul** [comment]
@otrosien I would like to, but I can't promise I'll have time for this anytime soon. (I still have #4898 on my plate, which I would like to finish first, and my time for Swagger-codegen is limited.)

**Comment by wing328** [comment]
@ePaul if you want, please feel free to work on this issue instead and leverage the community to work on #4898

**Comment by otrosien** [comment]
@ePaul I'm not too deep into the system property thing, but I assume that they were used to pass custom properties from the command line.

**Comment by ePaul** [comment]
Yeah, looks like it.

I did a short search for `System.getProperty()` over the codebase.

Property | used in | meaning
-------- | -------- | -------
swagger.codegen.inflector.apipackage | JavaInflectorServerCodegen | package name for API
swagger.codegen.inflector.modelpackage | JavaInflectorServerCodegen | package name for models
noservice | NodeJSServerCodegen | if not given, generate service.js from service.mustache.
swagger.codegen.undertow.apipackage | UndertowCodegen | package name for API
swagger.codegen.undertow.modelpackage | UndertowCodegen | package name for models
debugParser | DefaultCodegen | if given, output the parsed parameter structure given to `fromParameter`. (Was likely meant to be used in many more places instead of just this one.)
apis | DefaultGenerator | List of API names to generate (comma-separated)
models | DefaultGenerator | List of model names to generate (comma-separated)
supportingFiles | DefaultGenerator | List of names of supporting files to generate (comma-separated)
modelTests | DefaultGenerator | whether to generate tests for models (boolean, default true)
modelDocs | DefaultGenerator | whether to generate docs for models (boolean, default true)
apiTests | DefaultGenerator | whether to generate tests for APIs (boolean, default true)
apiDocs | DefaultGenerator | whether to generate docs for APIs (boolean, default true)
debugSwagger | DefaultGenerator | if set, output the `Swagger` object (as pretty-printed JSON) before starting the code generation.
debugModels | DefaultGenerator | if set, output the list of models (as pretty-printed JSON) at the end of generateModels
debugOperations | DefaultGenerator | if set, output the list of operations (as pretty-printed JSON) at the end of generateApis. Also, output some logging information at the start of processOperation.
debugSupportingFiles | DefaultGenerator | if set, output the created map (as pretty-printed JSON) at the end of buildSupportFileBundle

**Comment by ePaul** [comment]
So, the standard way of passing the model (and API) package names is via the additionalProperties' `modelPackage` (or `apiPackage`) entries. Those are taken in DefaultCodegen.processOpts into `setModelPackage`(`setApiPackage`).  (And CodegenConfigurator sets them into additionalProperties.)

This should work for JavaInflector and Undertow too. The system properties given just set default values which will be overwritten if a package name is given in the normal way.

**Comment by ePaul** [comment]
For the DefaultGenerator use cases, I'm going to add a `systemProperties` map into the OptClientInput class, which gets set by CodegenConfigurator. (PR coming up soon.)

**Comment by ePaul** [comment]
Just summing up my current idea of how this could/should work in the future:

* It is possible to use actual (real) system properties as input to Swagger-Codegen, i.e. when set from outside of the Java process.
* However, nothing in Swagger-Codegen itself will set system properties.
* CodegenConfigurator would take the (external) system properties and put them into its internal `systemProperties` map. The values of CLI's `-D` option as well as the maven-plugin's `<environmentVariables>` element go there, too (overriding them, if needed), but won't be copied into Java's system properties.
* The CliOptions can determine that some of those properties will get copied into the `additionalProperties` map.
* Everything inside Swagger-Codegen (the language codegens, DefaultGenerator, etc.) would read either from the `systemProperties` map or from `additionalProperties`, not from System.getProperty().

Any opinions?

I closed my PRs for this ticket, as I'll need to readjust them for how they fit into this strategy, and in which order to implement this to not break anything.
